### PR TITLE
Issue 51: Removed hardcoded oauth reference and added support for other registries and flows

### PR DESCRIPTION
**Pull request by jerrinss5**
Addressed issue 51 and 44. Made portieris generic to be interfaced with other registries.

### CODE REVIEW

File: helpers/oauth/header.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+	for c := 0; c < 256; c++ {
+		var t octetType
+		isCtl := c <= 31 || c == 127
+		isChar := 0 <= c && c <= 127
+		isSeparator := strings.ContainsRune(" \t\"(),/:;<=>?@[]\\{}", rune(c))
+		if strings.ContainsRune(" \t\r\n", rune(c)) {
+			t |= isSpace
+		}
+		if isChar && !isCtl && !isSeparator {
+			t |= isToken
```
Comment (molepigeon) [inline helpers/oauth/header.go]: I haven't seen an assignment like this in Go before - what does it do?

File: helpers/oauth/header.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+		}
+		octetTypes[c] = t
+	}
+}
+
+// ResponseChallenges takes in the response anf it is unauthorized returns a challenge containing realm and service information
```
Comment (molepigeon) [inline helpers/oauth/header.go]: typo: anf -> and

File: helpers/oauth/oauth.go [hunk 3]
```diff
@@ -44,34 +44,59 @@ var client = &http.Client{
...
-		"client_id":  {"testclient"},
-		"username":   {username},
-		"password":   {token},
-		"scope":      {"repository:" + repo + ":" + actions},
-	})
+	for _, challenge := range challengeSlice {
+		oauthEndpoint = challenge.Parameters["realm"]
+		service = challenge.Parameters["service"]
+		scope = challenge.Parameters["scope"]
+	}
```
Comment (molepigeon) [inline helpers/oauth/oauth.go]: If we had multiple challenges in the list, we'd end up with the realm, service, and scope from the last one. I think we'd only ever have one Challenge though, so I think that's okay... is that right?

File: helpers/oauth/header.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+	i := 0
+	for ; i < len(s); i++ {
+		if octetTypes[s[i]]&isToken == 0 {
+			break
+		}
+	}
+	return s[:i], s[i:]
+}
+
+func expectTokenOrQuoted(s string) (value, rest string) {
```
[the remark this answers is not in the document]
Comment (jerrinss5) [reply helpers/oauth/header.go]: I have referenced this from the docker's code here: https://github.com/docker/distribution/blob/master/registry/client/auth/challenge/authchallenge.go

File: helpers/oauth/header.go (hunk 1 above)
Comment (jerrinss5) [reply helpers/oauth/header.go]: it is like bitwise inclusive or and assign.
For example if var a is declared and unassigned and you a |= 4, value of 4 will be assigned.
https://play.golang.org/p/Hp46ShnG1x4
In the above code if `t` is unassigned then the value if isToken will be assigned to it.

Again I have referenced that entire header.go file from the docker's code - https://github.com/docker/distribution/blob/master/registry/client/auth/challenge/authchallenge.go

File: helpers/image/image.go [hunk 5]
```diff
@@ -150,6 +152,11 @@ func (r Reference) NameWithoutTag() string {
 	return r.name
 }
 
+// RepoName returns the image name without the tag.
+func (r Reference) RepoName() string {
+	return r.repo
+}
```
[the remark this answers is not in the document]
Comment (jerrinss5) [reply helpers/image/image.go]: it returns the details of the image minus the repo and tag.
NameWithoutTag() - returns image name without tag but contains the repo information.
eg: if the entire image name is Container Image: xyz.azurecr.io/notarysignerdemo/hello-world:signed2
NameWithoutTag() returns - xyz.azurecr.io/notarysignerdemo/hello-world
RepoName() returns - notarysignerdemo/hello-world

File: helpers/oauth/header.go (hunk 1 above)
Comment (molepigeon) [reply helpers/oauth/header.go]: Interesting, thanks!

File: helpers/image/image.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (molepigeon) [reply helpers/image/image.go]: Ok - so it looks to me that the difference is that `RepoName` doesn't include the server while `NameWithoutTag` does? If that's right, could we change the comment on this function to make clear that it doesn't include the domain?

[the remark this answers is not in the document]
Comment (jerrinss5) [reply helpers/image/image.go]: Sure i will update the comment accordingly.

**Comment by jerrinss5** [comment]
Thanks for the comments. I am working on it one by one, sorry would be a bit delayed addressing all of them, because of some personal commitments.

File: helpers/oauth/oauth.go [hunk 6]
```diff
@@ -44,34 +44,59 @@ var client = &http.Client{
...
+	if oauthEndpoint == "" || service == "" {
+		errMessage := "unable to fetch oauth realm and service header details"
+		glog.Errorf(errMessage)
+		return nil, fmt.Errorf(errMessage)
+	}
+
+	glog.Infof("Calling oauth endpoint: %s for registry service: %s and scope %s", oauthEndpoint, service, scope)
+	var resp *http.Response
+	var err error
+	if oauthEndpoint == "https://auth.docker.io/token" {
```
[the remark this answers is not in the document]
Comment (jerrinss5) [reply helpers/oauth/oauth.go]: so i was testing the POST call first to https://auth.docker.io/token but I got an http response code of 405. Not sure if I should code to handle both 404 and 405 out here? @molepigeon 
Log entries are as follows
`0922 23:05:29.855555       1 oauth.go:77] Calling oauth endpoint: https://auth.docker.io/token for registry service: notary.docker.io and scope repository:docker.io/library/nginx:pull
E0922 23:05:30.026918       1 oauth.go:96] Received non-success status code 405
E0922 23:05:30.027011       1 controller.go:200] Request to OAuth failed with status code: 405 and body:`

**Comment by chaosaffe** [comment]
@jerrinss5 are you still planning to work on this or should someone else pick up the PR?

**Comment by jerrinss5** [comment]
@chaosaffe - Lately I was not able to make much contribution to the PR. If someone would be able to help that would be great

File: helpers/image/image.go [hunk 7]
```diff
@@ -50,7 +51,7 @@ func NewReference(name string) (*Reference, error) {
 	}
 
 	// Get the hostname
-	hostname, _ := reference.SplitHostname(ref)
+	hostname, reponame := reference.SplitHostname(ref)
```
Comment (rubroboletus) [inline helpers/image/image.go]: SplitHostname is deprecated function, better to use latest github.com/docker/distribution/reference (v2.7.1+incompatible) and functions Domain and Path. Then you will get also the port.
hostname := reference.Domain(ref)
reponame := reference.Path(ref)

also the GetHostname() function is used only for querying notary and should contain port, if specified.
func (r Reference) GetHostname() string {
       port := r.port
       if port != "" {
           port = ":" + port
       }
       return r.hostname + port
}

Comment (chaosaffe) [reply helpers/image/image.go]: Thanks for the review @rubeoboletus! I will refactor this in #82

**Pull request closed without merging**